ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/select-media-devices/select-media-devices.component.ts

```diff
@@ -0,0 +1,151 @@
+    this.preferredMicrophoneStream = await this.userMediaStreamService
+      .getStreamForMic(preferredMicrophone);
+
+  }
+
+  async requestMedia() {
+    await this.userMediaStreamService.requestAccess();
+  }
+
+  private async initNewDeviceSelectionForm(): Promise<FormGroup> {
+    let cam = this.availableCameraDevices[0];
+    const preferredCamera = await this.userMediaService.getPreferredCamera();
+    if (preferredCamera) {
+      cam = this.availableCameraDevices.find(x => x.label === preferredCamera.label);
+    }
+
+    let mic = this.availableMicrophoneDevices[0];
+    const preferredMicrophone = await this.userMediaService.getPreferredMicrophone();
+    if (preferredMicrophone) {
+      mic = this.availableMicrophoneDevices.find(x => x.label === preferredMicrophone.label);
+    }
+    return this.formBuilder.group({
+      camera: [cam, Validators.required],
+      microphone: [mic, Validators.required]
+    });
+  }
+
+  get hasSingleCameraConncted(): boolean {
+    return this.availableCameraDevices.length === 1;
+  }
+
+  get singleCameraName(): string {
+    return this.availableCameraDevices[0].label;
+  }
+
+  get singleMicrophoneName(): string {
+    return this.availableMicrophoneDevices[0].label;
+  }
+
+  get hasSingleMicrophoneConncted(): boolean {
+    return this.availableMicrophoneDevices.length === 1;
+  }
+
+  get selectedCamera(): AbstractControl { return this.selectedMediaDevicesForm.get('camera'); }
+
+  get selectedMicrophone(): AbstractControl { return this.selectedMediaDevicesForm.get('microphone'); }
+
+  onSubmit() {
+    if (this.selectedMediaDevicesForm.invalid) {
+      return;
+    }
+    const selectedCam = this.getSelectedCamera();
+    const selectedMic = this.getSelectedMicrophone();
+    this.userMediaStreamService.stopStream(this.preferredCameraStream);
+    this.userMediaStreamService.stopStream(this.preferredMicrophoneStream);
+    this.acceptMediaDeviceChange.emit(new SelectedUserMediaDevice(selectedCam, selectedMic));
+  }
+
+  onCancel() {
+    this.cancelMediaDeviceChange.emit();
+  }
+
+  getSelectedCamera(): UserMediaDevice {
+    return this.availableCameraDevices.find(x => x === this.selectedCamera.value);
+  }
+
+  getSelectedMicrophone(): UserMediaDevice {
+    return this.availableMicrophoneDevices.find(x => x === this.selectedMicrophone.value);
+  }
+
+  private subscribeToDeviceSelectionChange() {
+    this.selectedCamera.valueChanges.subscribe(async newCamera => {
+      await this.updateCameraStream(newCamera);
+    });
+
+    this.selectedMicrophone.valueChanges.subscribe(async newMicrophone => {
+      await this.updateMicrophoneStream(newMicrophone);
+    });
+  }
+
+  private async updateCameraStream(newCam: UserMediaDevice) {
+    if (this.preferredCameraStream) {
+      this.userMediaStreamService.stopStream(this.preferredCameraStream);
+    }
+    this.userMediaService.updatePreferredCamera(newCam);
+    this.preferredCameraStream = null;
```

Comment:
Do you need to set to null, if next line reset it?

Reply:
@valentinabennett copied code from video web...